4337/contracts/Safe4337Module.sol

```diff
@@ -199,62 +182,55 @@ contract Safe4337Module is IAccount, HandlerContext, CompatibilityFallbackHandle
-        uint48 validUntil
-    ) internal view returns (bytes memory) {
-        bytes32 safeOperationHash = keccak256(
-            abi.encode(
-                SAFE_OP_TYPEHASH,
-                safe,
-                keccak256(callData),
-                nonce,
-                preVerificationGas,
-                verificationGasLimit,
-                callGasLimit,
-                maxFeePerGas,
-                maxPriorityFeePerGas,
-                validAfter,
-                validUntil,
-                SUPPORTED_ENTRYPOINT
-            )
-        );
-        return abi.encodePacked(bytes1(0x19), bytes1(0x01), domainSeparator(), safeOperationHash);
-    }
-
-    /**
-     * @dev Splits the user operation signature bytes into its parts.
-     * @param signatureData The user operation signature.
+     * @dev Decodes an ERC-4337 user operation into a Safe operation.
+     * @param userOp The ERC-4337 user operation.
+     * @return operationData Encoded EIP-712 Safe operation data bytes used for signature verification.
      * @return validAfter The timestamp the user operation is valid from.
      * @return validUntil The timestamp the user operation is valid until.
-     * @return signature The actual signature for the Safe user operation.
+     * @return signatures The Safe owner signatures extracted from the user operation.
      */
-    function _splitSignatureData(
-        bytes calldata signatureData
-    ) internal pure returns (uint48 validAfter, uint48 validUntil, bytes calldata signature) {
-        validAfter = uint48(bytes6(signatureData[0:6]));
-        validUntil = uint48(bytes6(signatureData[6:12]));
-        signature = signatureData[12:];
+    function _getSafeOp(
+        UserOperation calldata userOp
+    ) internal view returns (bytes memory operationData, uint48 validAfter, uint48 validUntil, bytes calldata signatures) {
+        // Extract additional Safe operation fields from the user operation signature which is encoded as:
+        // `abi.encodePacked(validAfter, validUntil, signatures)`
+        {
+            bytes calldata sig = userOp.signature;
+            validAfter = uint48(bytes6(sig[0:6]));
+            validUntil = uint48(bytes6(sig[6:12]));
+            signatures = sig[12:];
+        }
+
+        // It is important that **all** user operation fields are represented in the `SafeOp` data somehow, to prevent
+        // user operations from being submitted that do not fully respect the user preferences. The only exception is
+        // the `signature` bytes. Note that even `initCode` needs to be represented in the operation data, otherwise
+        // it can be replaced with a more expensive initialization that would charge the user additional fees.
+        {
+            // In order to work around Solidity "stack too deep" errors related to too many stack variables, manually
+            // encode the `SafeOp` fields into a memory `struct` for computing the EIP-712 struct-hash.
```

Comment:
Maybe it is worth mentioning that this works, because a struct that has no dynamic fields is represented as continuous memory in solidity. Therefore we can just pass it directly to keccak